### ISSUE TITLE
Catch UserExit

### ISSUE THREAD

**Issue by note-do** [oleg-shilo/wixsharp#429]
Hi, @oleg-shilo 
I want to develop the undo of after install actions if a user cancels the installation process. I use ManagedUI in my installer. I knew that in this case, wix# uses a workaround sending implied Managed Custom Action with `ActionResult.UserExit`. ManagedUI ExitDialog by default catches this state by `Shell.UserInterrupted` property.
How can I catch this state (user canceled installation) in the subscription of `event SetupEventHandler AfterInstall`? I have tried to get it from the `Shell.UserInterrupted`, but have got NPE because of this, I think:
`internal static ActionResult InvokeClientHandlers(Session session, string eventName, IShellView UIShell = null)`
Also, I've tried to use `ActionResult Result`, but I think it only related to the current event and have no information about the whole installation process. Could you get me some explanation for the definition of `SetupEventArgs`?

**Comment by oleg-shilo** [comment]
Unfortunately you cannot handle cancelled setup in the `AfterInstall` event. 

This is the MSI limitation. The `AfterInstall` event is invoked only after the install sequence is completed. The nature of installation is irrelevant. It can be fresh install, unisntall, upgrade or repair. 

But if the installation is interrupted, then the event handler is not invoked.

Remember, `AfterInstall` event handler is nothing else but a deferred custom action (CA). That's is why `ActionResult Result` is not suitable for you. This member is the user specified return value for that CA and it does not reflect the setup state but the desired result for CA (as you guessed).
Thus you can only sort of catch the cancellation from the ExitDialog.

**Comment by note-do** [comment]
Thank you. I think this workaround, doing the undo directly in `ExitDialog`, will be enough for me.
Is it possible to use `MsiRuntime.Data` in the ExitDialog?

**Comment by note-do** [comment]
I have researched the issue for a little. For some reasons the data that was set to the `SetupEventArgs.Data` in `AfterInstall ` subscription wasn't replicated in `MsiRuntime.Data`. Why is it happening?

**Comment by oleg-shilo** [comment]
It is indeed a very strange behavior.

There is very peculiar behavior when it comes to the synchronizing content of the `Session` object. 

At runtime MSI executes every custom action one by one. Interestingly enough every CA is executed as a separate process (rundll32.exe). MSI also pushes into CA a freshly initialized set of properties. After CA is completed MSI reads all properties from the session object and pushes them into the next CA.

EmbeddedUI is effectively an asynchronous long leaving CA. Unfortunately MSI never updates it when some other CA changed some properties. 

What even more inconvenient is that after the installation started (ProgressDialog) MSI disconnects the the session object so session properties cannot be accessed. Such a mess...

Anyway, I have created a test project (attached) that demonstrates the problem. It also demonstrates the solution. It shows how to put data to the session object in the first dialog and how to read it in the last dialog.

[WixSharp Setup3.zip](https://github.com/oleg-shilo/wixsharp/files/2201229/WixSharp.Setup3.zip)

**Comment by note-do** [comment]
> Anyway, I have created a test project (attached) that demonstrates the problem. It also demonstrates the solution. It shows how to put data to the session object in the first dialog and how to read it in the last dialog.

Unfortunately, the solution cannot help me in the case, as long as I need to push data from AfterInstall. But I'm really grateful to you for the brief explanation of the MSI session!

**Comment by oleg-shilo** [comment]
Not a problem, but keep in mind that AfterInstall event `Session` is also disconnected. I mean completely. It is because it is a _Deferred_ action. You cannot push anything into it at all.
MSI provided so called `CustomActionData` solution for tummeling Session properties from the normal session to the deferred CA. But this solution works only fr reading, not for writing. WixSharp does support this solution but it is obviously not your case.

Just another idea. Not elegant at all but it might just work. I use similar solution for detecting the setup cancellation:
- In the AfterInstall write to the log file `e.Session.Log("<special_tag>: my data");`
- In ExitDialog read the log file and parse the '<special_tag>' line.

This will work, but only if `e.Session.Log` possible from AfterInstall event.
